libsupport/include/katana/Cache.h

```diff
@@ -0,0 +1,157 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_CACHE_H_
+#define KATANA_LIBSUPPORT_KATANA_CACHE_H_
+
+// This is single threaded only, it is not intended to store large objects,
+// but rather metadata (e.g., a shared_ptr to a property column).
+
+// The problem witchel had implementing a multi-threaded version using
+// parallel-hashmap is a lock ordering problem.  parallel-hashmap 1.33 allows
+// execution of code with the write lock held, only when modifying, not when adding an
+// element.  We have to modify the LRU list when eviciting an element, so the natural
+// lock ordering is parallel-hashmap write lock, then list lock.  But without a way to
+// execute insert code with the parallel-hashmap write lock held, it seemed like there
+// would be some form of race condition.
+
+#include <list>
+#include <optional>
+#include <string>
+
+#include <boost/container_hash/hash.hpp>
+
+#include "katana/ArrowInterchange.h"
+#include "katana/Logging.h"
+
+namespace tsuba {
+
+template <typename Key, typename Value>
+class KATANA_EXPORT Cache {
+  using ListType = std::list<Key>;
+  struct MapValue {
+    Value value;
+    // This allows us to delete the old position in the LRU list without a scan
+    typename ListType::iterator lru_it;
+  };
+  using MapType = std::unordered_map<Key, MapValue>;
+
+public:
+  enum class ReplacementPolicy { kLRUSize, kLRUBytes };
+
+  Cache(
+      ReplacementPolicy policy = ReplacementPolicy::kLRUSize,
+      size_t lru_capacity = 0, size_t bytes_capacity = 0,
+      std::function<void(const Key& key)> evict_cb = nullptr,
+      std::function<size_t(const Value& value)> value_to_bytes = nullptr)
+      : policy_(policy),
+        lru_capacity_(lru_capacity),
+        bytes_capacity_(bytes_capacity),
+        evict_cb_(std::move(evict_cb)),
+        value_to_bytes_(std::move(value_to_bytes)) {
+    KATANA_LOG_VASSERT(
+        policy_ != ReplacementPolicy::kLRUSize || lru_capacity_ > 0,
+        "kLRUSize policy requires non-zero LRU entries");
+    KATANA_LOG_VASSERT(
+        policy_ != ReplacementPolicy::kLRUBytes || bytes_capacity_ > 0,
+        "kLRUBytes policy requires non-zero byte capacity");
+  }
+
+  size_t size() const { return key_to_value_.size(); }
+  size_t bytes() const { return total_bytes_; }
+
+  size_t LRUSizeCapacity() const { return lru_capacity_; }
```

Comment:
I don't know how I feel about having separate `size` and `bytes` and "capacity in size" and "capacity in bytes" methods. If  the Cache policy is an actual policy, the interface of Cache objects should not change when the policy changes. If there is reason for clients to care that they have a bytes-based Cache versus a item-based Cache, that indicates that the interface itself should be different (i.e., different classes) even if the implementation is largely similar.

Either:
- Make size and capacity report values in units appropriate for the cache policy (i.e., items for `kLRUSize` or bytes in `kLRUBytes`)
- Make two separate classes `LRUCache` and `LRUBytesCache` and optionally use private inheritance or composition to pull in the common implementation

Reply:
Gone with option 1.